src/fs/metadata_log_operations/unlink_or_remove_file.hh

```diff
@@ -0,0 +1,176 @@
+
+            auto entry_it = _dir_info->entries.find(_entry_name);
+            if (entry_it == _dir_info->entries.end()) {
+                return make_exception_future(no_such_file_or_directory_exception());
+            }
+            _entry_inode = entry_it->second;
+
+            _entry_inode_info = &_metadata_log._inodes.at(_entry_inode);
+            if (_entry_inode_info->is_directory()) {
+                if (not _remove_if_directory) {
+                    return make_exception_future(is_directory_exception());
+                }
+                if (not _entry_inode_info->get_directory().entries.empty()) {
+                    return make_exception_future(directory_not_empty_exception());
+                }
+            }
+
+            // Getting a lock on directory entry is enough to ensure it won't disappear because deleting directory
+            // requires it to be empty
+            if (_entry_inode_info->is_directory()) {
+                return _metadata_log._locks.with_locks(metadata_log::locks::unique {dir_inode, _entry_name}, metadata_log::locks::unique {_entry_inode}, [this] {
+                    return unlink_or_remove_file_in_directory();
+                });
+            } else {
+                return _metadata_log._locks.with_locks(metadata_log::locks::unique {dir_inode, _entry_name}, metadata_log::locks::shared {_entry_inode}, [this] {
+                    return unlink_or_remove_file_in_directory();
+                });
+            }
+        });
+    }
+
+    future<> unlink_or_remove_file_in_directory() {
+        if (not _metadata_log.inode_exists(_dir_inode)) {
+            return make_exception_future(operation_became_invalid_exception());
+        }
+
+        auto entry_it = _dir_info->entries.find(_entry_name);
+        if (entry_it == _dir_info->entries.end() or entry_it->second != _entry_inode) {
+            return make_exception_future(operation_became_invalid_exception());
+        }
+
+        if (_entry_inode_info->is_directory()) {
+            inode_info::directory& dir = _entry_inode_info->get_directory();
+            if (not dir.entries.empty()) {
+                return make_exception_future(directory_not_empty_exception());
+            }
+
+            assert(_entry_inode_info->directories_containing_file == 1);
+
+            // Ready to delete directory
+            ondisk_delete_inode_and_dir_entry_header ondisk_entry;
+            using entry_name_length_t = decltype(ondisk_entry.entry_name_length);
+            assert(_entry_name.size() <= std::numeric_limits<entry_name_length_t>::max());
+
+            ondisk_entry =  {
+                _entry_inode,
+                _dir_inode,
+                static_cast<entry_name_length_t>(_entry_name.size())
+            };
+
+            switch (_metadata_log.append_ondisk_entry(ondisk_entry, _entry_name.data())) {
+            case metadata_log::append_result::TOO_BIG:
+                assert(false and "ondisk entry cannot be too big");
```

Comment:
These kinds of errors also deserve throwing, not asserting. If we're unable to write an entry, there are graceful ways of shutting down the server, crashing it is too harsh.

Reply:
I think it is a bug that we cannot append entry to a fresh (empty) cluster -- it is too small to append this entry => it is too small for the file-system to function properly. So i treated it as a bug -- added assert for that. Maybe such checking (with exceptions) should be done in metadata_to_disk_buffer's constructor, but for me it is a bug here, where assert is situated.

Reply:
Ok, I'm fine with assertions as long as the input is properly validated earlier.

Reply:
I turns out this entry doesn't have fixed size, so indeed it should be an exception because we cannot check for it beforehand. Thx